### ISSUE TITLE
error: could not compile `pgrx-pg-sys` (lib)

### ISSUE THREAD

**Issue by dnlldl** [tensorchord/pgvecto.rs#197]
Hello again,
Now that I got through #193, I'm stuck a few steps later with the following error. I can't find any log or any way to get a more verbose output, any help would be appreciated to get a _better_ error message as this doesn't really say much. I did try to delete the whole toolchain folder, also tried running this as root (with or without the `sudo` parameter), tried most combinations of what I could think of but I still get stuck there.

The following output is after a 2nd run, as on a first run we can otherwise see the first 292/294 packages getting installed correctly until we get to `pgrx-pg-sys`:
```
postgres@vm-dev:/tmp/pgvecto.rs> cargo pgrx install --verbose -c /usr/lib/postgresql14/bin/pg_config --sudo --release
  2023-12-18T05:42:33.059603Z DEBUG cargo_pgrx::manifest: Found root package, manifest_path: /tmp/pgvecto.rs/Cargo.toml
    at cargo-pgrx/src/manifest.rs:64
    in cargo_pgrx::manifest::manifest_path
    in cargo_pgrx::command::install::execute

       Using PgConfig("pg14") and `pg_config` from /usr/lib/postgresql14/bin/pg_config
    Building extension with features pg14
     Running command "/var/lib/pgsql/.rustup/toolchains/nightly-2023-11-15-x86_64-unknown-linux-gnu/bin/cargo" "build" "--release" "--features" "pg14" "--no-default-features" "--message-format=json-render-diagnostics"
   Compiling pgrx-pg-sys v0.11.0 (https://github.com/tensorchord/pgrx.git?rev=7c30e2023876c1efce613756f5ec81f3ab05696b#7c30e202)
   Compiling pgrx v0.11.0 (https://github.com/tensorchord/pgrx.git?rev=7c30e2023876c1efce613756f5ec81f3ab05696b#7c30e202)
error: could not compile `pgrx-pg-sys` (lib)

Caused by:
  process didn't exit successfully: `/var/lib/pgsql/.rustup/toolchains/nightly-2023-11-15-x86_64-unknown-linux-gnu/bin/rustc --crate-name pgrx_pg_sys --edition=2021 /var/lib/pgsql/.cargo/git/checkouts/pgrx-82d5ad52313f9598/7c30e20/pgrx-pg-sys/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=237 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="pg14"' -C metadata=408fe13e8df11d13 -C extra-filename=-408fe13e8df11d13 --out-dir /tmp/pgvecto.rs/target/release/deps -L dependency=/tmp/pgvecto.rs/target/release/deps --extern libc=/tmp/pgvecto.rs/target/release/deps/liblibc-8e4be7041dd5ae5c.rmeta --extern memoffset=/tmp/pgvecto.rs/target/release/deps/libmemoffset-55dd90d72f818e08.rmeta --extern pgrx_macros=/tmp/pgvecto.rs/target/release/deps/libpgrx_macros-b7d1f1ad20b7ae6d.so --extern pgrx_sql_entity_graph=/tmp/pgvecto.rs/target/release/deps/libpgrx_sql_entity_graph-2e836f4cd8ebe45f.rmeta --extern serde=/tmp/pgvecto.rs/target/release/deps/libserde-4274cbb172fc7201.rmeta --extern sptr=/tmp/pgvecto.rs/target/release/deps/libsptr-085182e1b6e73607.rmeta --cap-lints allow --cfg nightly` (signal: 9, SIGKILL: kill)
```

**Comment by VoVAllen** [comment]
Ref: https://github.com/pgcentralfoundation/pgrx/issues/1103
Not sure about the root cause, it could be because of insufficient memory.

**Comment by dnlldl** [comment]
> Ref: [pgcentralfoundation/pgrx#1103](https://github.com/pgcentralfoundation/pgrx/issues/1103) Not sure about the root cause, it could be because of insufficient memory.

I noticed memory goes to 100% (6GB of RAM currently assigned), I'll try doubling the RAM and see what happens. Thanks for your help, I just wish the error messages were... more complete.

**Comment by dnlldl** [comment]
Doubled the memory, it did the trick. I'm now a little bit ahead but not too far lol

```
postgres@vm-dev:/tmp/pgvecto.rs> cargo pgrx install --verbose -c /usr/lib/postgresql14/bin/pg_config --sudo --release                 2023-12-18T06:30:20.654224Z DEBUG cargo_pgrx::manifest: Found root package, manifest_path: /tmp/pgvecto.rs/Cargo.toml
    at cargo-pgrx/src/manifest.rs:64
    in cargo_pgrx::manifest::manifest_path
    in cargo_pgrx::command::install::execute

       Using PgConfig("pg14") and `pg_config` from /usr/lib/postgresql14/bin/pg_config
    Building extension with features pg14
     Running command "/var/lib/pgsql/.rustup/toolchains/nightly-2023-11-15-x86_64-unknown-linux-gnu/bin/cargo" "build" "--release" "--features" "pg14" "--no-default-features" "--message-format=json-render-diagnostics"
   Compiling c v0.0.0 (/tmp/pgvecto.rs/crates/c)
   Compiling thiserror v1.0.50
The following warnings were emitted during compilation:

warning: c@0.0.0: error: unable to open output file '/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o': 'Operation not permitted'
warning: c@0.0.0: 1 error generated.

error: failed to run custom build command for `c v0.0.0 (/tmp/pgvecto.rs/crates/c)`

Caused by:
  process didn't exit successfully: `/tmp/pgvecto.rs/target/release/build/c-0b030a937f3b25a3/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=src/c.h
  cargo:rerun-if-changed=src/c.c
  TARGET = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CFLAGS_x86_64-unknown-linux-gnu
  CFLAGS_x86_64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CFLAGS_x86_64_unknown_linux_gnu
  CFLAGS_x86_64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=HOST_CFLAGS
  HOST_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  running: "/usr/bin/clang-16" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "--target=x86_64-unknown-linux-gnu" "-Wall" "-Wextra" "-o" "/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o" "-c" "./src/c.c"
  cargo:warning=error: unable to open output file '/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o': 'Operation not permitted'

  cargo:warning=1 error generated.

  exit status: 1

  --- stderr


  error occurred: Command "/usr/bin/clang-16" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "--target=x86_64-unknown-linux-gnu" "-Wall" "-Wextra" "-o" "/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o" "-c" "./src/c.c" with args "clang-16" did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...
```

clang is installed:
```
vm-dev:~ # clang --version
clang version 16.0.6
Target: x86_64-suse-linux
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by VoVAllen** [comment]
` unable to open output file '/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o': 'Operation not permitted'`
Sounds like a permission issue

**Comment by dnlldl** [comment]
> ` unable to open output file '/tmp/pgvecto.rs/target/release/build/c-cf1ce9345e397ac9/out/./src/c.o': 'Operation not permitted'` Sounds like a permission issue

Yup, it works as `root`. The guide could be slightly improved, I've never used Rust or Cargo before and honestly the most confusing parts for me were to figure out whether I was better to do everything as `root`, as `postgres` or as another user. I started doing with `root` but eventually got an error basically saying it couldn't be done as `root`, then tried everything as `postgres` but kept having to do some `chown -R` and have to deal with permissions.

Also as a non-initiated to Rust, it's hard to figure out how to undo stuff, I suspect most of my problems come from me failing to get through the procedure.

Anyways, I was able to re-execute as `postgres` as well after a `chown -R postgres:postgres .`, now I suspect there is some cleanup that could be done for `root` and maybe even for `postgres` (such as: that `/tmp/pgvecto.rs/` directory is probably not required anymore).
